cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/util/ProgramUtil.java

```diff
@@ -0,0 +1,287 @@
+  /**
+   * Json serializer/deserializer.
+   */
+  public static final Gson GSON = ApplicationSpecificationAdapter
+      .addTypeAdapters(new GsonBuilder())
+      .registerTypeAdapter(Trigger.class, new TriggerCodec())
+      .registerTypeAdapter(SatisfiableTrigger.class, new TriggerCodec())
+      .registerTypeAdapter(Constraint.class, new ConstraintCodec())
+      .create();
+
+  /**
+   * Json serde for decoding request. It uses a case insensitive enum adapter.
+   */
+  public static final Gson DECODE_GSON = ApplicationSpecificationAdapter
+      .addTypeAdapters(new GsonBuilder())
+      .registerTypeAdapterFactory(new CaseInsensitiveEnumTypeAdapterFactory())
+      .registerTypeAdapter(Trigger.class, new TriggerCodec())
+      .registerTypeAdapter(SatisfiableTrigger.class, new TriggerCodec())
+      .registerTypeAdapter(Constraint.class, new ConstraintCodec())
+      .create();
+
+  public static <T extends BatchProgram> List<T> validateAndGetBatchInput(FullHttpRequest request,
+      Type type)
+      throws BadRequestException, IOException {
+
+    List<T> programs;
+    try (Reader reader = new InputStreamReader(new ByteBufInputStream(request.content()),
+        StandardCharsets.UTF_8)) {
+      try {
+        programs = DECODE_GSON.fromJson(reader, type);
+        if (programs == null) {
+          throw new BadRequestException(
+              "Request body is invalid json, please check that it is a json array.");
+        }
+      } catch (JsonSyntaxException e) {
+        throw new BadRequestException("Request body is invalid json: " + e.getMessage());
+      }
+    }
+
+    // validate input
+    for (BatchProgram program : programs) {
+      try {
+        program.validate();
+      } catch (IllegalArgumentException e) {
+        throw new BadRequestException(
+            "Must provide valid appId, programType, and programId for each object: "
+                + e.getMessage());
+      }
+    }
+    return programs;
+  }
+
+  /**
+   * Parses the give program type into {@link ProgramType} object.
+   *
+   * @param programType the program type to parse.
+   *
+   * @throws BadRequestException if the given program type is not a valid {@link ProgramType}.
+   */
+  public static ProgramType getProgramType(String programType) throws BadRequestException {
+    try {
+      return ProgramType.valueOfCategoryName(programType);
+    } catch (Exception e) {
+      throw new BadRequestException(String.format("Invalid program type '%s'", programType), e);
+    }
+  }
+
+  /**
+   * Returns the program status with no need of application existence check.
+   *
+   * @param appSpec the ApplicationSpecification of the existing application
+   * @param programId the id of the program for which the status call is made
+   * @return the status of the program
+   * @throws NotFoundException if the application to which this program belongs was not found
+   */
+  public ProgramStatus getExistingAppProgramStatus(ApplicationSpecification appSpec,
```

Comment:
From what  I see, this guy and all callers are only used from ProgramLicecycleService. Let's have it in there
![image](https://github.com/user-attachments/assets/cf346548-51d8-4f90-a0a4-ac6b92a767e4)

Reply:
Great suggestion! Done.